File: data/lipsofsuna/core/specs/terrain-material.lua
--- Specifies the behavior of terrain materials.
--
-- Lips of Suna is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Lesser General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.
--
-- @module core.specs.terrain_material
-- @alias TerrainMaterialSpec

local Class = require("system/class")
local Spec = require("core/specs/spec")

--- Specifies the behavior of terrain materials.
-- @type TerrainMaterial
local TerrainMaterialSpec = Spec:register("TerrainMaterialSpec", "terrain material", {
	{name = "name", type = "string", description = "Name of the spec."},
	{name = "categories", type = "dict", dict = {type = "boolean"}, default = {}, description = "Dictionary of categories."},
	{name = "effect_build", type = "string", description = "Building effect.", details = {spec = "EffectSpec"}},
	{name = "effect_collapse", type = "string", description = "Collapsing effect.", details = {spec = "EffectSpec"}},
	{name = "mining_item", type = "string", description = "Item name obtained by mining.", details = {spec = "ItemSpec"}},
	{name = "texture", type = "string", description = "Default texture for all faces."},
	{name = "texture_top", type = "string", description = "Texture name for the top surface"},
	{name = "texture_bottom", type = "string", description = "Texture name for the bottom surface"},
	{name = "texture_side", type = "string", description = "Texture name for the side surfaces"},
	{name = "texture_decoration", type = "string", description = "Texture name of the decoration"},
	{name = "decoration_type", type = "number", description = "Decoration type"},
	{name = "stick_type", type = "number", description = "Stick type"}
})

--- Creates a new terrain material specification.
-- @param clss TerrainMaterialSpec class.
-- @param args Arguments.
-- @return TerrainMaterialSpec.
TerrainMaterialSpec.new = function(clss, args)
	local self = Spec.new(clss, args)
	return self
end

return TerrainMaterialSpec

File: src/lipsofsuna/extension/terrain/terrain-material.c
/* Lips of Suna
 * Copyright© 2007-2013 Lips of Suna development team.
 *
 * Lips of Suna is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Lips of Suna is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Lips of Suna. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * \addtogroup LIExt Extension
 * @{
 * \addtogroup LIExtTerrain Terrain
 * @{
 */

#include "terrain-material.h"

/**
 * \brief Initializes the material.
 * \param self Material.
 * \param texture Texture number.
 */
void liext_terrain_material_init (
	LIExtTerrainMaterial* self,
	int                   texture)
{
	self->stick_type = LIEXT_TERRAIN_STICK_TYPE_DEFAULT;
	self->decoration_type = LIEXT_TERRAIN_DECORATION_TYPE_NONE;
	self->texture_top = texture;
	self->texture_bottom = texture;
	self->texture_side = texture;
	self->texture_decoration = texture;
	self->texture_scaling = 0.2f;
}

/** @} */
/** @} */
